packages/terra-date-picker/src/DateInput.jsx

```diff
@@ -91,139 +117,535 @@ const propTypes = {
+    value,
+    ...customProps
+  } = props;
 
-    this.handleOnButtonClick = this.handleOnButtonClick.bind(this);
-    this.handleOnChange = this.handleOnChange.bind(this);
-    this.handleOnKeyDown = this.handleOnKeyDown.bind(this);
-  }
+  const [date, dateDispatch] = useReducer(dateReducer, { day: '', month: '', year: '' });
+  const [isFocused, setFocused] = useState(false);
+  const [dayInitialFocused, setDayInitialFocused] = useState(false);
+  const [monthInitialFocused, setMonthInitialFocused] = useState(false);
+  const [yearInitialFocused, setYearInitialFocused] = useState(false);
+  const dayInputRef = useRef();
+  const monthInputRef = useRef();
+  const yearInputRef = useRef();
+  const uuidRef = useRef(uuidv4());
+  const theme = React.useContext(ThemeContext);
+
+  const { onCalendarButtonClick, shouldShowPicker } = customProps;
+  delete customProps.onCalendarButtonClick;
+  delete customProps.shouldShowPicker;
+
+  const additionalInputProps = { ...customProps, ...inputAttributes };
+  const momentDateFormat = useMemo(() => DateUtil.getFormatByLocale(intl.locale), [intl.locale]);
+  const dateValue = DateUtil.convertToISO8601(value, momentDateFormat);
+  const dateFormatOrder = DateUtil.getDateFormatOrder(momentDateFormat);
+  const separator = DateUtil.getDateSeparator(intl.locale);
+
+  const setDayRef = useCallback(node => {
+    dayInputRef.current = node;
 
-  componentDidUpdate(prevProps) {
-    if (this.shouldShowPicker && !prevProps.shouldShowPicker && this.props.onClick) {
-      this.props.onClick();
+    // Logic used internally by date-time-picker to shift focus to time-input from day input when date order is YMD.
+    if (inputRefCallback && dateFormatOrder === DateUtil.dateOrder.YMD) {
+      inputRefCallback(node);
     }
-  }
+  }, [dateFormatOrder, inputRefCallback]);
+
+  const setMonthRef = useCallback(node => {
+    monthInputRef.current = node;
+  }, []);
 
-  handleOnButtonClick(event) {
-    const attributes = { ...this.props.inputAttributes };
+  const setYearRef = useCallback(node => {
+    yearInputRef.current = node;
 
-    if (!attributes.readOnly && this.onCalendarButtonClick && this.props.onClick) {
-      this.onCalendarButtonClick(event, this.props.onClick);
+    // Logic used internally by date-time-picker to shift focus to time-input from year input when date order is MDY or DMY.
+    if (inputRefCallback && (dateFormatOrder === DateUtil.dateOrder.MDY || dateFormatOrder === DateUtil.dateOrder.DMY)) {
+      inputRefCallback(node);
     }
-  }
+  }, [dateFormatOrder, inputRefCallback]);
 
-  handleOnChange(event) {
-    if (!DateUtil.validDateInput(event.target.value)) {
-      return;
+  // Triggers the onClick callback to launch the dropdown picker for the scenario when the default date is invalid and
+  // the calendar button is clicked which should clear the value and launch the dropdown picker
+  useEffect(() => {
+    if (shouldShowPicker && onClick) {
+      onClick();
```

Comment:
I'm a little confused about invoking a click action as an effect. How is this being used? With the current implementation any changes to `shouldShowPicker` will invoke a click callback, is this correct?

Reply:
I guess @benbcai could give better insight on this.